ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,536 @@
+use std::{borrow::Borrow, ops::Sub};
+
+use futures_util::{stream, StreamExt};
+use generic_array::{ArrayLength, GenericArray};
+use typenum::{Unsigned, U1};
+
+use crate::{
+    ff::PrimeField,
+    helpers::Direction,
+    protocol::{
+        context::{
+            dzkp_field::{BlockSize, UVTupleBlock},
+            Context,
+        },
+        ipa_prf::malicious_security::{
+            lagrange::{CanonicalLagrangeDenominator, LagrangeTable},
+            prover::{ProofArray, UVPolynomial, UVStore, ZeroKnowledgeProof},
+        },
+        prss::SharedRandomness,
+        RecordId,
+    },
+    seq_join::seq_join,
+};
+
+/// This is a batch of `ZeroKnowledgeProofs` owned by a verifier.
+/// It consists of a `ProofBatch` from the prover on the left
+/// and a `ProofBatch` from the prover on the right.
+///
+/// It uses parameter `R` which is the recursion factor of the zero-knowledge proofs
+/// A zero-knowledge proof with recursion factor `R` has length `2*R::USIZE-1`
+///
+/// The `VerifierBatch` also contains two masks `p_0` and `q_0` in `masks` stored as `(p_0,q_0)`
+/// These masks are used as additional `u,v` values for the final proof
+/// of the `left` batch of `ZeroKnowledgeProofs`.
+/// On the `right` batch of `ZeroKnowledgeProofs` these values are set to `F::ZERO`.
+/// These masks mask sensitive information when the right verifier sends a field value to the left verifier,
+/// i.e. using a channel with `Direction:Right`, who does not have access to these masks.
```

Comment:
Yes, this will happen later during the verification. The **verifier** on the right will send information to the **verifier** on the left using `Direction::Right`. This is why we need the masks. During the proof generation, we use `Direction::Left` to send the proof shares from the **prover** to the verifier on the left.